File: DSA(Geeks for Geeks)/Basics/Array/Lectures/05. Reverse an array.js
// Reverse an array

/**
 * I/P: arr = [10, 5, 7, 30]
 * O/P: arr = [30, 7, 5, 10]
 * 
 * I/P: arr = [30, 20, 10]
 * O/P: arr = [10, 20, 30]
 * 
 * I/P: arr = [10]
 * O/P: arr = [10]
 */

// ================================
// Naive Solution
/**
 * 1. Create a temp array
 *      temp = []
 * 2. Traverse the input array from right to left and append elements to temp.
 *      temp = [30, 20, 10]
 * 
 * 3. Copy temp elements back to arr.
 *      arr = [30, 20, 10]
 * 
 * --> Time complexity: O(n)
 * --> Space complexity: O(n)
 */

// ==================================
// Efficient Solution

function reverseArr(arr) {
    let low = 0;
    let high = arr.length - 1;
    while(low < high) {
        [arr[low], arr[high]] = [arr[high], arr[low]];
        low++;
        high--;
    }
    return arr;
}

console.log(reverseArr([10, 5, 3]))

File: python/DSA(Geeks for Geeks)/Basics/Array/05. Reverse an Array.py
# // Reverse an array

# /**
#  * I/P: arr = [10, 5, 7, 30]
#  * O/P: arr = [30, 7, 5, 10]
#  * 
#  * I/P: arr = [30, 20, 10]
#  * O/P: arr = [10, 20, 30]
#  * 
#  * I/P: arr = [10]
#  * O/P: arr = [10]
#  */

def reverse (arr) :
    low = 0
    high = len(arr) - 1

    while(low < high) :
        temp = arr[low]
        arr[low] = arr[high]
        arr[high] = temp
        low = low + 1
        high = high - 1

    return arr

print(reverse([10, 5, 7, 30]))

# ============================================

def reverseAlt (arr) :
    low = 0
    high = len(arr) - 1

    while(low < high) :
        arr[low] , arr[high] = arr[high] , arr[low]
        low = low + 1
        high = high - 1

    return arr

print(reverseAlt([10, 5, 7, 30]))